### PR TITLE
Add `SyncBatchNormPlugin`

### LINKED ISSUE #4349: Support sync batch norm as a plugin
## 🚀 Feature
Support sync batch norm as a plugin so we don't duplicate code across accelerators

### Motivation

It'll look very similar to this: https://github.com/PyTorchLightning/pytorch-lightning/pull/4285

### Pitch

```
import torch
from pytorch_lightning.core.lightning import LightningModule


class SyncBatchNormPlugin(object):
    """
    Plugin to link a custom sync batchnorm implementation to any arbitrary accelerator.
    Example::
        class MySyncBN(SyncBatchNormPlugin):

            def configure_sync_batchnorm(self, model):
                model = MySyncBNWrapper(model)
                return model
        my_sync_bn = MySyncBN()
        trainer = Trainer(sync_batchnorm=True, plugins=[my_sync_bn])
    """

    def configure_sync_batchnorm(self, model: LightningModule) -> LightningModule:
        """
        By default: adds global batchnorm for a model spread across multiple GPUs and nodes.
        Override this to synchronize batchnorm between specific process groups instead
        of the whole world or use a different sync_bn like `apex`'s version.
        Args:
            model: pointer to current :class:`LightningModule`.
        Return:
            LightningModule with batchnorm layers synchronized between process groups
        """
        model = torch.nn.SyncBatchNorm.convert_sync_batchnorm(model, process_group=None)

        return model
```

Do we even need the `sync_batchnorm` flag on the `Trainer` with this plugin?

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #4349

- Added a simple base class for plugins that wrap layers for syncing during forward/backward
- Added a implementation for native sync batch norm
- Added tests for all cases of configuration

This decouples the baked-in support for sync batchnorm in ddp- and ddp spawn strategies. This would for example enable adding a plugin for Apex-specific sync batch norm in the future.

TODO: 
- [x] Tests
- [x] Rebase onto #11448 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda @akihironitta

### CODE REVIEW

**Review by ananthsub — COMMENTED** [review]
Back when I filed the issue, my reasoning to make this a plugin was to support both pytorch native & Apex: https://nvidia.github.io/apex/_modules/apex/parallel.html#convert_syncbn_model

Almost all instances I've seen have moved to the native version. But i'm sure there are a handful of users who still might use the apex version. Making this an interface will make it easier to support apex later. That could be a good first issue

**Comment by awaelchli** [comment]
Ah interesting! Then we should probably aim for an abstract interface and a concrete default torch implementation. I'll explore it, thanks!

File: pytorch_lightning/plugins/sync_batchnorm.py [hunk 1]
```diff
@@ -19,7 +20,23 @@
...
+    @abstractmethod
+    def apply(model: "pl.LightningModule") -> "pl.LightningModule":
+        """Override this static method to apply synchronization to the layers of this model."""
+
+    @staticmethod
+    @abstractmethod
+    def revert(model: "pl.LightningModule") -> "pl.LightningModule":
+        """Override this static method to undo all modifications made in :meth:`apply`."""
+
+
```
Comment (ananthsub) [inline pytorch_lightning/plugins/sync_batchnorm.py]: - i don't think these have to be static methods. someone could provide specialized instances. the strategy would still call self.layer_sync.apply/revert

- making these not nn.Module based means these could be applied to Lite
```suggestion
    @abstractmethod
    def apply(self, model: nn.Module) -> nn.Module:
        """Override this static method to apply synchronization to the layers of this model."""

    @abstractmethod
    def revert(self, nn.Module) -> nn.Module:
        """Override this static method to undo all modifications made in :meth:`apply`."""


```

File: pytorch_lightning/plugins/sync_batchnorm.py [hunk 2]
```diff
@@ -19,7 +20,23 @@
...
+class NativeSyncBatchNorm(ABC):
```
Comment (ananthsub) [inline pytorch_lightning/plugins/sync_batchnorm.py]:
```suggestion
class NativeSyncBatchNorm(LayerSync):
```

File: tests/models/test_sync_batchnorm.py [hunk 3]
```diff
@@ -1,131 +0,0 @@
...
-        loss = F.cross_entropy(y_hat, y)
-
-        return loss
-
-    def configure_optimizers(self):
-        return torch.optim.Adam(self.linear.parameters(), lr=0.02)
-
-
-# TODO: Fatal Python error: Bus error
-@pytest.mark.skip(reason="Fatal Python error: Bus error")
```
Comment (awaelchli) [inline tests/models/test_sync_batchnorm.py:72]: this test was skipped and outdated (failing for other reasons)
I rewrote it and moved it to parity tests

File: pytorch_lightning/utilities/distributed.py [hunk 4]
```diff
@@ -367,40 +366,6 @@ def _collect_states_on_rank_zero(state: Dict[str, Any]) -> Dict[int, Any]:
     return {rank: _broadcast_object_list(state, rank) for rank in range(torch.distributed.get_world_size())}
 
 
-class _BatchNormXd(torch.nn.modules.batchnorm._BatchNorm):
```
Comment (awaelchli) [inline pytorch_lightning/utilities/distributed.py:370]: moved these utilities without modification

File: pytorch_lightning/plugins/sync_batchnorm.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/sync_batchnorm.py]: done

File: tests/strategies/test_custom_plugin.py [hunk 5]
```diff
@@ -18,27 +18,8 @@
...
-
-class CustomParallelStrategy(DDPStrategy):
-    def __init__(self, **kwargs):
-        super().__init__(**kwargs)
-        # Set to None so it will be overwritten by the accelerator connector.
-        self.sync_batchnorm = None
-
-
-@RunIf(skip_windows=True)
-def test_sync_batchnorm_set(tmpdir):
```
Comment (awaelchli) [inline tests/strategies/test_custom_plugin.py:34]: moved to `test_accelerator_connector` where it belongs

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 6]
```diff
@@ -291,6 +294,13 @@ def _check_config_and_set_final_flags(
                     self.checkpoint_io = plugin
                 elif isinstance(plugin, ClusterEnvironment):
                     self._cluster_environment_flag = plugin
+                elif isinstance(plugin, LayerSync):
+                    if sync_batchnorm and not isinstance(plugin, NativeSyncBatchNorm):
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:298]: Shouldn't this check the base interface?

```suggestion
                    if sync_batchnorm and not isinstance(plugin, LayerSync):
```

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 7]
```diff
@@ -291,6 +294,13 @@ def _check_config_and_set_final_flags(
...
+                        raise MisconfigurationException(
+                            "You set `Trainer(sync_batchnorm=True)` and provided a `LayerSync` plugin, but this"
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:300]:
```suggestion
                            "You set `Trainer(sync_batchnorm=True)` and provided a `{plugin.__class__.__name__}` plugin, but this"
```

File: pytorch_lightning/plugins/layer_sync.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    def revert(self, model: Module) -> Module:
+        """Convert the wrapped batchnorm layers back to regular batchnorm layers.
+
+        Args:
+            model: Reference to the current LightningModule
+
+        Return:
+            LightningModule with regular batchnorm layers that will no longer sync across processes.
+        """
+        return _revert_sync_batchnorm(model)
```
Comment (carmocca) [inline pytorch_lightning/plugins/layer_sync.py:64]: I saw that you moved them but why not move the `_revert_sync_batchnorm` directly inside of this method?

File: pytorch_lightning/strategies/ddp.py [hunk 9]
```diff
@@ -86,7 +87,7 @@ def __init__(
         )
         log.detail(f"{self.__class__.__name__}: initializing DDP plugin")
         self._num_nodes = 1
-        self.sync_batchnorm = False
+        self._layer_sync: Optional[LayerSync] = None
```
Comment (carmocca) [inline pytorch_lightning/strategies/ddp.py:90]: Should this be defined instead in `ParallelStrategy.__init__`?

**Review by carmocca — APPROVED** [review]
Looking great!

File: pytorch_lightning/plugins/layer_sync.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+
+import torch
+from torch.nn import Module
+
+
+class LayerSync(ABC):
```
Comment (carmocca) [inline pytorch_lightning/plugins/layer_sync.py:21]: Have you considered making this a contextmanager even if we don't use it with a simple `with` statement?

File: pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 6 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:298]: no because I want to express here that `Trainer(sync_batchnorm=True, plugins= NativeSyncBatchNorm()) `is valid and should not error, while any other conflicting choice would create an error.

File: pytorch_lightning/plugins/layer_sync.py (hunk 8 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/layer_sync.py:64]: Because we will have the apex version for sync batchnorm and will probably want to share code.

File: pytorch_lightning/plugins/layer_sync.py (hunk 10 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/layer_sync.py:21]: it requires the model as input. i don't see how a context manager could possibly access the model(s) instances under it.

Comment (carmocca) [reply pytorch_lightning/plugins/layer_sync.py:21]: Something like this?

```python
class LayerSync:
    def __init__(self, model):
        self.model = model  # weakref.proxy?

    def __enter__(self) -> Module:
        pass

    def __exit__(self, exc_type, exc_value, exc_traceback) -> Module:
        pass

    def apply(self) -> Module:
        return self.__enter__()

    def revert(self) -> Module:
        return self.__exit__(None, None, None)
```

File: pytorch_lightning/plugins/layer_sync.py (hunk 8 above)
Comment (carmocca) [reply pytorch_lightning/plugins/layer_sync.py:64]: I suggest that we move them outside when we want to share it, just in case it doesn't work out later.

File: pytorch_lightning/plugins/layer_sync.py (hunk 10 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/layer_sync.py:21]: I still don't see this. It has no effect on the code under the context manager. What is the meaning of this?

```py
with LayerSync(model):
    model(x)
    model2(x)
```

Let's not do this. Even if it somehow made sense, nobody can't take advantage of this "feature". Also, we don't want people using plugins outside the trainer. That will only limit us later when we want to develop the plugin further.

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 11]
```diff
@@ -317,6 +320,13 @@ def _check_config_and_set_final_flags(
                     self.checkpoint_io = plugin
                 elif isinstance(plugin, ClusterEnvironment):
                     self._cluster_environment_flag = plugin
+                elif isinstance(plugin, LayerSync):
+                    if sync_batchnorm and not isinstance(plugin, NativeSyncBatchNorm):
+                        raise MisconfigurationException(
+                            f"You set `Trainer(sync_batchnorm=True)` and provided a `{plugin.__class__.__name__}`"
+                            f" plugin, but this is not allowed. Choose one or the other."
```
Comment (kaushikb11) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:327]:
```suggestion
                            " plugin, but this is not allowed. Choose one or the other."
```

**Review by kaushikb11 — APPROVED** [review]
Awesome!

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/plugins/layer_sync.py [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+    @abstractmethod
+    def apply(self, model: Module) -> Module:
+        """Override this method to apply synchronization to the layers of this model."""
+
+    @abstractmethod
+    def revert(self, model: Module) -> Module:
+        """Override this method to undo all modifications made in :meth:`apply`."""
+
+
+class NativeSyncBatchNorm(LayerSync):
```
Comment (justusschock) [inline pytorch_lightning/plugins/layer_sync.py:34]: So we don't support the Apex variant of it? I know that we didn't so far, but now that we have an interface for this, we could :)

File: pytorch_lightning/strategies/parallel.py [hunk 13]
```diff
@@ -42,6 +43,7 @@ def __init__(
         super().__init__(accelerator=accelerator, checkpoint_io=checkpoint_io, precision_plugin=precision_plugin)
         self.parallel_devices = parallel_devices
         self.cluster_environment = cluster_environment
+        self._layer_sync: Optional[LayerSync] = None
```
Comment (justusschock) [inline pytorch_lightning/strategies/parallel.py:46]: Just a question: Why is this plugin private if other plugins (like cluster env and precision) are not?

**Review by justusschock — APPROVED** [review]
LGTM, 2 very minor questions :)

File: pytorch_lightning/plugins/layer_sync.py (hunk 12 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/layer_sync.py:34]: Yes, it's a motivation of this feature. I would be opening a feature issue after we merge this.

File: pytorch_lightning/strategies/parallel.py (hunk 13 above)
Comment (awaelchli) [reply pytorch_lightning/strategies/parallel.py:46]: I think mainly as a security because we set it from the accelerator connector. We may also want to expose it as a constructor arg in the future because the others are too. Just trying to give us max flexibility after 1.6 when we declare it as stable. We can always add new optional args and properties :)

**Review by tchaton — APPROVED** [review]
LGTM !
